### PR TITLE
cleanup and enhancement of JEDEC SPI NOR support

### PR DESCRIPTION
**Pull request by pabigot**
Fixes bad unit addresses in the bindings used to test littlefs.

Changes `jedec,spi-nor` binding to require the `size` property, which is in fact required for the driver to build.  Adds the property to all declared compatible nodes in-tree from which it was missing.

Extends `jedec,spi-nor` binding to identify support for deep-power-down mode, a common feature that can reduce standby mode current consumption by as much as 99.9% (MX25L32 from 10 uA to 3 uA; MX24R64 from 9 uA to 7 nA).

Updates driver to place chip into DPD when it is not active.

### CODE REVIEW

File: dts/bindings/mtd/jedec,spi-nor.yaml [hunk 1]
```diff
@@ -19,11 +19,71 @@ properties:
...
     type: boolean
     required: false
-    description: Indicates the device supports the BE32K command
+    description: Indicates the device supports the BE32K (0xD8) command
 
-  size:
+  requires-ulbpr:
+    type: boolean
+    required: false
+    description: Indicates the device requires the ULBPR (0x98) command
```
Comment (ulfalizer) [inline dts/bindings/mtd/jedec,spi-nor.yaml]: Could maybe use one of the multi-line syntaxes in https://yaml-multiline.info/ (just discovered that site... pretty specialized) and put a period at the end of the first lines.

Comment (ulfalizer) [reply dts/bindings/mtd/jedec,spi-nor.yaml]: Tell me if the output for these long descriptions gets ugly btw. Could improve it in that case.

Comment (pabigot) [reply dts/bindings/mtd/jedec,spi-nor.yaml]: I'd looked at that site, but it didn't seem to add anything beyond what's normally used.  I think I've used `>` in previous bindings; `>-` doesn't add anything unless the value is actually used in generated documentation somewhere.  Is it?

The vast majority of description text in `dts/bindings` does not have a terminating period.

Comment (ulfalizer) [reply dts/bindings/mtd/jedec,spi-nor.yaml]: Could skip the period if it's meant to be some kind of title though I guess...

Comment (ulfalizer) [reply dts/bindings/mtd/jedec,spi-nor.yaml]: > I'd looked at that site, but it didn't seem to add anything beyond what's normally used.

Nice to have the first line line up with the other lines I think.

> I think I've used `>` in previous bindings; `>-` doesn't add anything unless the value is actually used in generated documentation somewhere. Is it?

Should end up in the generarated header/configuration file. Think it's just `rstrip()`ed at the moment.

Comment (ulfalizer) [reply dts/bindings/mtd/jedec,spi-nor.yaml]: Without looking, I suspect that it gets a bit ugly.

I'll make a PR tomorrow so that it gets output as

```c
/*
 * Long description
 * yada yada
 * yada
 */
```

for multi-line descriptions and as

```c
/* yada yada */
```

for single-line descriptions.

File: drivers/flash/spi_nor.c [hunk 2]
```diff
@@ -166,10 +227,14 @@ static int spi_nor_read(struct device *dev, off_t addr, void *dest,
 
 	SYNC_LOCK();
 
+	(void)exit_dpd(dev);
```
Comment (vanwinkeljan) [inline drivers/flash/spi_nor.c]: Better implement  device power management for the driver instead of putting the device in low power mode for each transaction and leave the decision to put the device in low power mode to the application code.

File: drivers/flash/spi_nor.c [hunk 3]
```diff
@@ -314,6 +391,7 @@ static int spi_nor_configure(struct device *dev)
 	const struct spi_nor_config *params = dev->config->config_info;
 
 	data->spi = device_get_binding(DT_INST_0_JEDEC_SPI_NOR_BUS_NAME);
+	printk("Dev %s at %p\n", DT_INST_0_JEDEC_SPI_NOR_BUS_NAME, data->spi);
```
Comment (vanwinkeljan) [inline drivers/flash/spi_nor.c]: Use logging infrastructure instead of printk.

**Review by vanwinkeljan — CHANGES_REQUESTED** [review]

File: drivers/flash/spi_nor.c (hunk 3 above)
Comment (pabigot) [reply drivers/flash/spi_nor.c]: Left-over debug message, will remove.

File: drivers/flash/spi_nor.c (hunk 2 above)
Comment (pabigot) [reply drivers/flash/spi_nor.c]: I'd love to, but it's useless.  There's no way to propagate the dependency information down to the driver unless you're interacting with the flash directly.

File: dts/bindings/mtd/jedec,spi-nor.yaml (hunk 1 above)
Comment (pabigot) [reply dts/bindings/mtd/jedec,spi-nor.yaml]: I think all these yaml issues have been addressed.

File: drivers/flash/spi_nor.c (hunk 2 above)
Comment (pabigot) [reply drivers/flash/spi_nor.c]: @vanwinkeljan Please re-check this.

I've added a Kconfig option to allow control of whether deep power-down is entered between transactions.  This should preserve existing behavior, while still allowing my use case of dropping current consumption in applications which rarely use the flash.

If you absolutely insist I will go ahead and implement the device power management (eventually), but as I note in the commit message and above it's not every useful until we get device dependency in place so that clients of the device can ensure it's adequately woken before it's used.  As it stands many applications won't even know (or care) whether their file system is on a SPI NOR device, or how to access that device to control its power consumption.

Comment (vanwinkeljan) [reply drivers/flash/spi_nor.c]: Thx for the update, adding the Kconfig is fine for me.

**Review by vanwinkeljan — APPROVED** [review]

**Review by gon1332 — APPROVED** [review]

**Comment by pabigot** [comment]
Trivial style and comment updates, and addition of a check that the configured flash page layout size is a value that can be erased.  Lack of such a check may have contributed to #19438.

```
drivers: spi_nor: improve documentation and validation on flash layout size

The smallest region that can be erashed is one sector, so setting
SPI_NOR_FLASH_LAYOUT_PAGE_SIZE=2048 will fail at runtime when the
flash page API is used to erase a single (or misaligned) page.  Add a
compile-time check that the requested layout page size is erasable.
```

File: boards/arm/mimxrt1015_evk/mimxrt1015_evk.dts [hunk 4]
```diff
@@ -53,6 +53,7 @@ arduino_serial: &uart4 {};
 	reg = <0x402a8000 0x4000>, <0x60000000 0x1000000>;
 	at25sf128a: at25sf128a@0 {
 		compatible = "adesto,at25sf128a", "jedec,spi-nor";
+		size = <134217728>;
```
Comment (nvlsianpu) [inline boards/arm/mimxrt1015_evk/mimxrt1015_evk.dts]: Isn't it possible to put number in hexadecimal representation here? I believe it will by natural.

File: dts/bindings/mtd/jedec,spi-nor.yaml [hunk 5]
```diff
@@ -24,13 +24,68 @@ properties:
...
+
+  dpd-wakeup-sequence:
+    type: array
+    required: false
+    description: |
+      Specifies wakeup durations for devices without RDPD.
+
+      Some devices (Macronix MX25R in particular) wake from deep power
+      down by a timed sequence of CSn toggles rather than the RDPD
+      command.  This property specifies three durations measured in
```
Comment (nvlsianpu) [inline dts/bindings/mtd/jedec,spi-nor.yaml]: Can you specify here which time is which in DT property sequence? `<x y z>`

**Review by nvlsianpu — COMMENTED** [review]
will comment more soon

File: boards/arm/mimxrt1015_evk/mimxrt1015_evk.dts (hunk 4 above)
Comment (pabigot) [reply boards/arm/mimxrt1015_evk/mimxrt1015_evk.dts]: It really isn't, though that's what I used to do.  The datasheets tend to provide them in decimal, and for normal humans it's a lot easier to understand that `134217728` is 128 Mbit and `67108864` is 64 Mbit than to try to understand what `0x8000000` and `0x400000` mean.

Also the majority of nodes that weren't added by me use decimal, so I decided to follow the common path.

File: dts/bindings/mtd/jedec,spi-nor.yaml (hunk 5 above)
Comment (pabigot) [reply dts/bindings/mtd/jedec,spi-nor.yaml]: I thought it was clear, but maybe the update is better.

**Comment by nvlsianpu** [comment]
Added the `enhancement` label as the patch introduces power-save feature for JEDEC SPI NOR devices.

File: drivers/flash/spi_nor.c [hunk 6]
```diff
@@ -110,6 +135,87 @@ static int spi_nor_access(const struct device *const dev,
...
+{
+	int ret = 0;
+
+	if (IS_ENABLED(DT_INST_0_JEDEC_SPI_NOR_HAS_DPD)) {
+#ifdef DT_INST_0_JEDEC_SPI_NOR_DPD_WAKEUP_SEQUENCE
+		/* Delay for tDPDD */
+		k_sleep(K_MSEC(T_DPDD_MS));
+		/* Assert CSn and wait for tCRDP.
+		 *
+		 * Unfortunately the SPI API doesn't allow us to
```
[the remark this answers is not in the document]
Comment (pabigot) [reply drivers/flash/spi_nor.c]: Yes.  This is certainly shaky, since depending on the delay and the bus frequency the timing requirement might not be met, but without control of CSn directly I don't have an alternative.  (We can't just grab the GPIO and muck with it because some drivers don't provide CSn GPIO information.)  It's also possible but unlikely that sending the command would cause the chip to misbehave.  I could see substituting `RDID` instead.

Experimentally this works on at least two devices that have this (efr32mg_sltb002a's MX25R80 support isn't enabled in tree yet, but I've tested it).

File: drivers/flash/spi_nor.c [hunk 7]
```diff
@@ -110,6 +135,87 @@ static int spi_nor_access(const struct device *const dev,
...
+}
+
+static int exit_dpd(const struct device *const dev)
+{
+	int ret = 0;
+
+	if (IS_ENABLED(DT_INST_0_JEDEC_SPI_NOR_HAS_DPD)) {
+#ifdef DT_INST_0_JEDEC_SPI_NOR_DPD_WAKEUP_SEQUENCE
+		/* Delay for tDPDD */
+		k_sleep(K_MSEC(T_DPDD_MS));
```
[the remark this answers is not in the document]
Comment (pabigot) [reply drivers/flash/spi_nor.c]: Yeah, ok, we could do that.  I'd have to capture the uptime clock in `enter_dpd` then do some calculations here.  Off-hand that seems doable.  I'll wait until the complete set of requested changes is in place before tackling it, though.

File: boards/arm/mimxrt1015_evk/mimxrt1015_evk.dts (hunk 4 above)
Comment (nvlsianpu) [reply boards/arm/mimxrt1015_evk/mimxrt1015_evk.dts]: I can accept that, better align to majority in this case

File: drivers/flash/spi_nor.c (hunk 7 above)
[the remark this answers is not in the document]
Comment (pabigot) [reply drivers/flash/spi_nor.c]: I've added support to capture the time at which DPD was entered, and delay where necessary before exiting.  This also allowed eliminating the unconditional wait for enter DPD to complete before proceeding.

File: drivers/flash/spi_nor.c (hunk 6 above)
[the remark this answers is not in the document]
Comment (pabigot) [reply drivers/flash/spi_nor.c]: Switched to using `RDID`, just in case.

File: dts/bindings/mtd/jedec,spi-nor.yaml [hunk 8]
```diff
@@ -24,13 +24,68 @@ properties:
   requires-ulbpr:
```
[the remark this answers is not in the document]
Comment (nvlsianpu) [reply dts/bindings/mtd/jedec,spi-nor.yaml]: satisfied

**Review by nvlsianpu — COMMENTED** [review]
Looks good now, I will approve once CI become green.

**Comment by pabigot** [comment]
> Looks good now, I will approve once CI become green.

Finally green (rebased and pushed to clear the stalled tasks).

**Review by nvlsianpu — APPROVED** [review]

**Comment by nvlsianpu** [comment]
@ulfalizer Can you revisit?

File: samples/drivers/spi_flash/sample.yaml [hunk 9]
```diff
@@ -16,3 +16,10 @@ tests:
             - "Data read 55 aa"
             - "Data read matches with data written. Good!!"
     depends_on: spi
+  sample.driver.spi_flash_dpd:
+    tags: spi flash
+    filter: DT_INST_0_JEDEC_SPI_NOR_LABEL
```
Comment (galak) [inline samples/drivers/spi_flash/sample.yaml]: This `filter:` needs updating, we removed support for `DT_` symbol based filters, need to use a function like `filter: dt_compat_enabled("jedec,spi-nor")`

File: drivers/flash/spi_nor.c [hunk 10]
```diff
@@ -110,6 +185,88 @@ static int spi_nor_access(const struct device *const dev,
...
+ *
+ * This means taking the lock and, if necessary, waking the device
+ * from deep power-down mode.
+ */
+static void acquire_device(struct device *dev)
+{
+#ifdef CONFIG_MULTITHREADING
+	struct spi_nor_data *const driver_data = dev->driver_data;
+
+	k_sem_take(&driver_data->sem, K_FOREVER);
```
Comment (tbursztyka) [inline drivers/flash/spi_nor.c]: sem attribute seem to be always present, so you could use IS_ENABLED(CONFIG_MULTITHREAD) here

File: drivers/flash/spi_nor.c [hunk 11]
```diff
@@ -110,6 +185,88 @@ static int spi_nor_access(const struct device *const dev,
...
+static void release_device(struct device *dev)
+{
+	if (IS_ENABLED(CONFIG_SPI_NOR_IDLE_IN_DPD)) {
+		enter_dpd(dev);
+	}
+
+#ifdef CONFIG_MULTITHREADING
+	struct spi_nor_data *const driver_data = dev->driver_data;
+
+	k_sem_give(&driver_data->sem);
```
Comment (tbursztyka) [inline drivers/flash/spi_nor.c]: same here

File: drivers/flash/spi_nor.c [hunk 12]
```diff
@@ -33,19 +69,58 @@ struct spi_nor_data {
...
+#ifdef DT_INST_0_JEDEC_SPI_NOR_HAS_DPD
+	/* Low 32-bits of uptime counter at which device last entered
+	 * deep power-down.
+	 */
+	u32_t ts_enter_dpd;
+#endif
 	struct k_sem sem;
 };
 
-#if defined(CONFIG_MULTITHREADING)
```
Comment (tbursztyka) [inline drivers/flash/spi_nor.c]: actually, this SYNC_* macros being replaced by new functions should be in its own commit. It is not related to deep power down.

Comment (pabigot) [reply drivers/flash/spi_nor.c]: It sorta is, since it's the added complexity of handling DPD that makes the macros unwieldy.  However, if you really want it that way I can do it.

File: boards/arm/mimxrt1015_evk/mimxrt1015_evk.dts (hunk 4 above)
Comment (galak) [inline boards/arm/mimxrt1015_evk/mimxrt1015_evk.dts]: its a nit, but for readability can we use DT_SIZE_M

File: boards/arm/mimxrt1020_evk/mimxrt1020_evk.dts [hunk 13]
```diff
@@ -61,6 +61,7 @@ arduino_serial: &uart2 {};
 	reg = <0x402a8000 0x4000>, <0x60000000 0x800000>;
 	is25wp064: is25wp064@0 {
 		compatible = "issi,is25wp064", "jedec,spi-nor";
+		size = <67108864>;
```
Comment (galak) [inline boards/arm/mimxrt1020_evk/mimxrt1020_evk.dts]: its a nit, but for readability can we use DT_SIZE_M

File: boards/arm/mimxrt1050_evk/mimxrt1050_evk_qspi.dts [hunk 14]
```diff
@@ -12,6 +12,7 @@
 	reg = <0x402a8000 0x4000>, <0x60000000 0x800000>;
 	is25wp064: is25wp064@0 {
 		compatible = "issi,is25wp064", "jedec,spi-nor";
+		size = <67108864>;
```
Comment (galak) [inline boards/arm/mimxrt1050_evk/mimxrt1050_evk_qspi.dts]: its a nit, but for readability can we use DT_SIZE_M

File: boards/arm/mimxrt1060_evk/mimxrt1060_evk.dts [hunk 15]
```diff
@@ -68,6 +68,7 @@ arduino_serial: &uart3 {};
 	reg = <0x402a8000 0x4000>, <0x60000000 0x800000>;
 	is25wp064: is25wp064@0 {
 		compatible = "issi,is25wp064", "jedec,spi-nor";
+		size = <67108864>;
```
Comment (galak) [inline boards/arm/mimxrt1060_evk/mimxrt1060_evk.dts]: its a nit, but for readability can we use DT_SIZE_M

File: boards/riscv/hifive1/hifive1.dts [hunk 16]
```diff
@@ -63,6 +63,7 @@
 	reg = <0x10014000 0x1000 0x20400000 0xc00000>;
 	flash0: flash@0 {
 		compatible = "issi,is25lp128", "jedec,spi-nor";
+		size = <134217728>;
```
Comment (galak) [inline boards/riscv/hifive1/hifive1.dts]: its a nit, but for readability can we use DT_SIZE_M

File: boards/riscv/hifive1_revb/hifive1_revb.dts [hunk 17]
```diff
@@ -61,6 +61,7 @@
 	reg = <0x10014000 0x1000 0x20010000 0x3c0900>;
 	flash0: flash@0 {
 		compatible = "issi,is25lp128", "jedec,spi-nor";
+		size = <134217728>;
```
Comment (galak) [inline boards/riscv/hifive1_revb/hifive1_revb.dts]: its a nit, but for readability can we use DT_SIZE_M

File: boards/riscv/qemu_riscv32/qemu_riscv32.dts [hunk 18]
```diff
@@ -39,6 +39,7 @@
 	reg = <0x10014000 0x1000 0x20400000 0xc00000>;
 	flash0: flash@0 {
 		compatible = "issi,is25lp128", "jedec,spi-nor";
+		size = <134217728>;
```
Comment (galak) [inline boards/riscv/qemu_riscv32/qemu_riscv32.dts]: its a nit, but for readability can we use DT_SIZE_M

File: boards/riscv/qemu_riscv64/qemu_riscv64.dts [hunk 19]
```diff
@@ -43,6 +43,7 @@
 	reg = <0x10014000 0x1000 0x20400000 0xc00000>;
 	flash0: flash@0 {
 		compatible = "issi,is25lp128", "jedec,spi-nor";
+		size = <134217728>;
```
Comment (galak) [inline boards/riscv/qemu_riscv64/qemu_riscv64.dts]: its a nit, but for readability can we use DT_SIZE_M

**Review by galak — COMMENTED** [review]
Looks good, minor nit about using DT_SIZE_M macro to make things a bit more readable.

File: boards/arm/mimxrt1015_evk/mimxrt1015_evk.dts (hunk 4 above)
Comment (pabigot) [reply boards/arm/mimxrt1015_evk/mimxrt1015_evk.dts]: Could be confusing, since the size is measured in bits.  This one would have to be `DT_SIZE_M(128 * 8)`.  Also DT_SIZE_M appears to be Zephyr-specific, which isn't ideal.

Do you want to require this?  For consistency we should also change the existing `size` properties in other devicetree files.

Comment (galak) [reply boards/arm/mimxrt1015_evk/mimxrt1015_evk.dts]: Ah, didn't realize size was in bits... (just curious, is it ever not a multiple of 8?).

Comment (pabigot) [reply boards/arm/mimxrt1015_evk/mimxrt1015_evk.dts]: It's never not a multiple of 4096 * 8 since the sector size is always 4096 bytes.  Almost certainly never not a multiple of 65536 * 8, which is the standard block erase size (some chips support 32768 byte block erases).

In early days I used the hex representation, but switched to decimal because it was what most other people were using, and because it's easier for a human to recognize the size.  Few people can look at 0x8000000 and know immediately that it's a 128 Mbit (16 MByte) flash.

Comment (pabigot) [reply boards/arm/mimxrt1015_evk/mimxrt1015_evk.dts]: Also, to look back at history: when I converted to DTS I made the size bits instead of bytes because the datasheets always specify the size in bits rather than bytes.  It's common in some applications to use multiple flash devices in parallel: bits makes more sense in that situation.

File: drivers/flash/spi_nor.c [hunk 20]
```diff
@@ -348,11 +490,19 @@ static int spi_nor_configure(struct device *dev)
 	data->spi_cfg.cs = &data->cs_ctrl;
 #endif /* DT_INST_0_JEDEC_SPI_NOR_CS_GPIOS_CONTROLLER */
 
+	/* Might be in DPD if system restarted without power cycle. */
+	(void)exit_dpd(dev);
```
Comment (tbursztyka) [inline drivers/flash/spi_nor.c]: If you add such (void), maybe you'll need it everywhere where relevant?

**Review by tbursztyka — APPROVED** [review]
fine by me. just a tiny comment

File: drivers/flash/spi_nor.c (hunk 20 above)
Comment (pabigot) [reply drivers/flash/spi_nor.c]: Easier to remove it than try to explain why it makes more sense here than other places, so it's gone.  Thanks for the approval.

Comment (tbursztyka) [reply drivers/flash/spi_nor.c]: ok, I was also wondering why not just removing it. We do not enforce returned code check everywhere after all

File: boards/arm/mimxrt1015_evk/mimxrt1015_evk.dts (hunk 4 above)
Comment (galak) [reply boards/arm/mimxrt1015_evk/mimxrt1015_evk.dts]: Can we do something like:

`size = <134217728>; /* in bits */`

File: drivers/flash/spi_nor.c (hunk 20 above)
Comment (pabigot) [reply drivers/flash/spi_nor.c]: OK, so the reason I did it is that here I really don't care if it fails.  We've already confirmed that we can communicate with the device, and we only need to do this to handle a case where a soft reset with DPD active switched to code that doesn't enable DPD when idle.

The draft Zephyr coding standard does require checking return values everywhere (though the estimate of 186 violations in the code base seems ridiculously low) so all this may have to change anyway.  (Depending on what functions are deemed to "return error information".)

File: boards/arm/mimxrt1015_evk/mimxrt1015_evk.dts (hunk 4 above)
Comment (pabigot) [reply boards/arm/mimxrt1015_evk/mimxrt1015_evk.dts]: Of course we could.  But why?  The property description in the binding specifies that the value is in bits.  A wrong value will be discovered quite rapidly.

This sort of thing is a slippery slope where for consistency we'd end up going through and adding a units value to every numeric device tree property that has units (timeout, other sizes, ...).  I don't really think that's a bad idea, but I also don't think it should be done ad hoc.
